In [1]:
import numpy as np
import itertools

**1) Как относятся друг к другу множество и последовательность? (в ответе использовать слова типа: часть, целое, общее, частное, родитель, дочерний субъект и т.д.)?**

Последовательность это функция аргументом которой является элемент из множества натуральных чисел. Значения этой функции принадлежат некоторому множеству, которое называется областью значения функции.

**2) Прочитать высказывания математической логики, построить их отрицания и *установить истинность.***


$\forall y \in[0 ; 1] : \operatorname{sgn}(y)=1$

Для любого y, принадлежащего отрезку от 0 до 1 включительно, верно, что значение функции sgn(y) равно 1.  
**Это выражение ложно.**  

Отрицание:  
$\exists y \in[0 ; 1] : \operatorname{sgn}(y)\not=1$

$\forall n \in \mathbb{N}>2 : \exists x, y, z \in \mathbb{N} : x^{n}=y^{n}+z^{n}$

Для любого n из множества натуральных чисел и больше 2 существуют такие x,y,z из множества натуральных чисел, что справедливо равенство x^n = y^n + z^n  
**Это выражение ложно.**  

Отрицание:  
$\exists n \in \mathbb{N}>2 : \forall x, y, z \in \mathbb{N} : x^{n}\not=y^{n}+z^{n}$


$\forall x \in \mathbb{R} \exists X \in \mathbb{R} : X>x$

Для любого x из множества действительных чисел существует такое X из множества действительных чисел, что X > x.  
**Это выражение истинно.**  

Отрицание:  
$\exists x \in \mathbb{R} \forall X \in \mathbb{R} : X\leq x$


$\forall x \in \mathbb{C} \not \exists y \in \mathbb{C} : x>y \| x<y$

Для любого х из множества комплексных чисел не существует такого y из множества комплексных чисел что справедливо неравенство x > y или x < y  
**Это выражение ложно, так как нельзя сравнивать комплексные числа на отношение больше/меньше.**  

Отрицание:  
$\exists x \in \mathbb{C} \exists y \in \mathbb{C} : x \leq y , x \geq y$

$\forall y \in\left[0 ; \frac{\pi}{2}\right] \exists \varepsilon>0 : \sin y<\sin (y+\varepsilon)$

Для любого y, принадлежащего отрезку от 0 до $\frac{\pi}2$ включительно, существует такое положительное значение $\varepsilon$, что справедливо неравенство sin(y) < sin(y+$\varepsilon$)  
**Это выражение ложно.**  

Отрицание:  
$\exists y \in\left[0 ; \frac{\pi}{2}\right] \forall \varepsilon>0 : \sin y \geq \sin (y+\varepsilon)$

$\forall y \in[0 ; \pi) \exists \varepsilon>0 : \cos y>\cos (y+\varepsilon)$

Для любого y, принадлежащего отрезку от 0 включительно до ${\pi}$ ислючительно, существует такое положительное значение $\varepsilon$, что справедливо неравенство cos(y) > cos(y+$\varepsilon$)  
**Это выражение ложно.**  

Отрицание:  
$\exists y \in[0 ; \pi) \forall \varepsilon>0 : \cos y \leq \cos (y+\varepsilon)$


$\exists x : x \notin\{\mathbb{N}, \mathbb{Z}, \mathbb{Q}, \mathbb{R}, \mathbb{C}\}$

Для любого х справедливо, что х принадлежит множеству натуральных или целых или рациональных или вещественных или комплексных чисел.  
**Это варажение ложно.**  
Отрицание:  
$\forall x : x \in\{\mathbb{C}\}$


**3.1) Даны три множества a,b и с. Необходимо выполнить все изученные виды бинарных операций над всеми комбинациями множеств.**  
**3.2) *Выполнить задание 3.1 на языке Python**

###### Памятка для себя
**Бинарные операции**:  
1) Пересечение $A \cap B$ (A.intersection(B))  
2) Объединение $A \cup B$ (A.union(B))   
3) Разность $A \backslash B$ (A.difference(B))  
4) Симметрическая разность $A \triangle B$ (A.symmetric_difference(B))  
5) Декартово произведение $A \times B$ (def dekart_product(A,B): import itertools; return set(itertools.product(*[A,B])))  
    
**Унарные операции**:  
1) Дополнение  
2) Булеан

**Порядок выполнения:**  
В первую очередь выполняются унарные операции, затем пересечение, затем все остальные (слева направо), имеющие равный приоритет.

In [2]:
A = set([1,2,3,4])
B = set([2,4,6,8,10])
C = set([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11])
setlist = [A, B, C]

In [3]:
binary_ops = [("intersection", 0, u'\u22C2'), 
              ("union", 1, u'\u22C3'), 
              ("subtraction", 2, "\\"), 
              ("symm_diff", 3, u'\u25B3'), 
              ("dek_prod", 4, u'\u2A2F')]

In [4]:
def dekart_product(A,B):
    return set(itertools.product(*[A,B]))

In [5]:
def solve_bin_ops(A, B, op_code):
#    ("intersection", 0, u'\u22C2')
#    ("union", 1, u'\u22C3')
#    ("subtraction", 2, "\\")
#    ("symm_diff", 3, u'\u25B3')
#    ("dek_prod", 4, u'\u2A2F')
    if op_code == 0:
        return A.intersection(B)
    elif op_code == 1:
        return A.union(B)
    elif op_code == 2:
        return A.difference(B)
    elif op_code == 3:
        return A.symmetric_difference(B)
    elif op_code == 4:
        return dekart_product(A, B)
    else:
        print("Wrong op_code:", op_code)

In [6]:
def solve_expression(sets_, ops_):
    if len(sets_) == 3:
        op1 = ops_[0][1]
        op2 = ops_[1][1]
        if op2 == 0 and op1 != 0:
            summary = solve_bin_ops(sets_[1], sets_[2], op2)
            summary = solve_bin_ops(sets_[0], summary, op1)
        else:
            summary = solve_bin_ops(sets_[0], sets_[1], op1)
            summary = solve_bin_ops(summary, sets_[2], op2)
        return summary
    elif len(sets_) == 2:
        op = ops_[0][1]
        return solve_bin_ops(sets_[0], sets_[1], op)
    else:
        print("Wrong sets count")

In [7]:
def namestr(obj, namespace=globals()):
    for name in namespace:
        if namespace[name] is obj:
            return name

In [8]:
def parse_combination(combination_):
    return [(namestr(comb_val), comb_val) for comb_val in combination_]

In [9]:
three_sets = itertools.product(setlist, repeat=3)
two_sets = itertools.product(setlist, repeat=2)
two_ops = list(itertools.product(binary_ops, repeat=2))
one_op = list(itertools.combinations(binary_ops, 1))

In [10]:
for combination_ in two_sets:
    parsed_combination = parse_combination(combination_)
    set_names = [val[0] for val in parsed_combination]
    sets_ = [val[1] for val in parsed_combination]
    for op in one_op:
        print(set_names[0], op[0][2], set_names[1], "=", solve_expression(sets_, op))
        print()

A ⋂ A = {1, 2, 3, 4}

A ⋃ A = {1, 2, 3, 4}

A \ A = set()

A △ A = set()

A ⨯ A = {(1, 2), (3, 2), (1, 3), (4, 1), (3, 3), (3, 1), (4, 4), (2, 1), (1, 4), (2, 4), (2, 3), (4, 3), (2, 2), (4, 2), (3, 4), (1, 1)}

A ⋂ B = {2, 4}

A ⋃ B = {1, 2, 3, 4, 6, 8, 10}

A \ B = {1, 3}

A △ B = {1, 3, 6, 8, 10}

A ⨯ B = {(4, 8), (2, 8), (1, 6), (1, 2), (3, 10), (4, 4), (3, 6), (2, 2), (1, 10), (4, 10), (3, 2), (2, 6), (1, 4), (2, 10), (4, 2), (4, 6), (3, 8), (1, 8), (3, 4), (2, 4)}

A ⋂ C = {1, 2, 3, 4}

A ⋃ C = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}

A \ C = set()

A △ C = {5, 6, 7, 8, 9, 10, 11}

A ⨯ C = {(4, 7), (1, 3), (4, 8), (2, 8), (3, 11), (2, 1), (1, 6), (3, 7), (2, 5), (1, 11), (1, 2), (4, 9), (3, 3), (2, 9), (3, 10), (4, 4), (1, 5), (3, 6), (2, 2), (1, 10), (4, 1), (1, 1), (4, 10), (3, 2), (2, 6), (4, 5), (1, 4), (2, 10), (3, 9), (2, 3), (1, 9), (4, 2), (4, 11), (3, 5), (2, 7), (4, 6), (3, 1), (2, 11), (3, 8), (1, 8), (4, 3), (1, 7), (3, 4), (2, 4)}

B ⋂ A = {2, 4}

B ⋃ A = {1, 2, 3, 4, 6,

In [11]:
for combination_ in three_sets:
    parsed_combination = parse_combination(combination_)
    set_names = [val[0] for val in parsed_combination]
    sets_ = [val[1] for val in parsed_combination]
    for ops in two_ops:
        print(set_names[0], ops[0][2], set_names[1], ops[1][2], set_names[2], "=", solve_expression(sets_, ops))
        print()

A ⋂ A ⋂ A = {1, 2, 3, 4}

A ⋂ A ⋃ A = {1, 2, 3, 4}

A ⋂ A \ A = set()

A ⋂ A △ A = set()

A ⋂ A ⨯ A = {(1, 2), (3, 2), (1, 3), (4, 1), (3, 3), (3, 1), (4, 4), (2, 1), (1, 4), (2, 4), (2, 3), (4, 3), (2, 2), (4, 2), (3, 4), (1, 1)}

A ⋃ A ⋂ A = {1, 2, 3, 4}

A ⋃ A ⋃ A = {1, 2, 3, 4}

A ⋃ A \ A = set()

A ⋃ A △ A = set()

A ⋃ A ⨯ A = {(1, 2), (3, 2), (1, 3), (4, 1), (3, 3), (3, 1), (4, 4), (2, 1), (1, 4), (2, 4), (2, 3), (4, 3), (2, 2), (4, 2), (3, 4), (1, 1)}

A \ A ⋂ A = set()

A \ A ⋃ A = {1, 2, 3, 4}

A \ A \ A = set()

A \ A △ A = {1, 2, 3, 4}

A \ A ⨯ A = set()

A △ A ⋂ A = set()

A △ A ⋃ A = {1, 2, 3, 4}

A △ A \ A = set()

A △ A △ A = {1, 2, 3, 4}

A △ A ⨯ A = set()

A ⨯ A ⋂ A = {(1, 2), (3, 2), (1, 3), (4, 1), (3, 3), (3, 1), (4, 4), (2, 1), (1, 4), (2, 4), (2, 3), (4, 3), (2, 2), (4, 2), (3, 4), (1, 1)}

A ⨯ A ⋃ A = {1, (3, 2), (1, 3), 2, 3, 4, (2, 1), (1, 4), (1, 1), (2, 3), (4, 2), (1, 2), (3, 3), (3, 1), (4, 4), (4, 3), (2, 2), (3, 4), (4, 1), (2, 4)}

A ⨯ A \ A = {(1, 2), (

A ⨯ C ⨯ C = {((1, 4), 6), ((1, 10), 3), ((2, 5), 11), ((3, 7), 1), ((4, 2), 3), ((3, 4), 3), ((4, 11), 11), ((2, 8), 11), ((1, 6), 9), ((1, 5), 3), ((4, 5), 8), ((3, 9), 1), ((3, 10), 7), ((2, 2), 11), ((2, 7), 6), ((1, 2), 11), ((1, 3), 9), ((4, 10), 6), ((2, 11), 4), ((1, 4), 9), ((2, 10), 5), ((4, 9), 1), ((1, 10), 10), ((3, 7), 10), ((4, 2), 4), ((2, 9), 11), ((4, 6), 6), ((1, 5), 4), ((1, 1), 6), ((3, 8), 10), ((3, 9), 6), ((2, 4), 8), ((3, 5), 4), ((1, 9), 4), ((3, 6), 8), ((4, 3), 10), ((1, 2), 2), ((2, 11), 3), ((2, 10), 2), ((4, 9), 8), ((1, 7), 10), ((2, 6), 1), ((3, 2), 11), ((2, 9), 2), ((4, 8), 10), ((1, 11), 5), ((2, 5), 7), ((4, 1), 5), ((3, 8), 1), ((2, 4), 1), ((1, 6), 5), ((1, 8), 11), ((3, 10), 9), ((3, 6), 3), ((3, 1), 2), ((4, 7), 11), ((3, 11), 5), ((2, 1), 7), ((2, 10), 11), ((1, 7), 3), ((2, 7), 10), ((2, 6), 6), ((3, 2), 2), ((4, 10), 10), ((2, 9), 5), ((3, 3), 3), ((4, 6), 8), ((4, 8), 3), ((3, 4), 6), ((3, 5), 10), ((4, 11), 6), ((1, 5), 8), ((4, 5), 3), ((1,

B △ A ⋃ C = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}

B △ A \ C = set()

B △ A △ C = {2, 4, 5, 7, 9, 11}

B △ A ⨯ C = {(6, 9), (1, 3), (10, 11), (6, 6), (10, 6), (3, 11), (8, 9), (6, 2), (1, 6), (3, 7), (1, 11), (8, 5), (10, 3), (1, 2), (10, 8), (6, 7), (3, 3), (8, 1), (10, 7), (3, 10), (6, 10), (10, 2), (8, 10), (6, 3), (1, 5), (3, 6), (1, 10), (8, 6), (1, 1), (10, 9), (6, 4), (3, 2), (8, 2), (10, 4), (6, 11), (8, 11), (1, 4), (3, 9), (1, 9), (8, 7), (10, 1), (6, 5), (3, 5), (8, 3), (10, 5), (6, 8), (10, 10), (6, 1), (3, 1), (3, 8), (1, 8), (8, 8), (1, 7), (3, 4), (8, 4)}

B ⨯ A ⋂ C = {(2, 1), (6, 2), (8, 4), (10, 3), (8, 1), (4, 4), (6, 3), (2, 2), (4, 1), (6, 4), (8, 2), (10, 4), (2, 3), (10, 1), (4, 2), (8, 3), (6, 1), (4, 3), (2, 4), (10, 2)}

B ⨯ A ⋃ C = {(6, 4), 1, 2, 3, (8, 2), (10, 4), 4, 5, 6, 7, 8, 9, 10, (2, 1), 11, (6, 2), (2, 3), (10, 1), (4, 2), (10, 3), (8, 3), (8, 1), (6, 1), (4, 4), (10, 2), (6, 3), (4, 3), (2, 2), (4, 1), (2, 4), (8, 4)}

B ⨯ A \ C = {(2, 1), (6, 2), (10,

C △ A ⋃ B = {2, 4, 5, 6, 7, 8, 9, 10, 11}

C △ A \ B = {9, 11, 5, 7}

C △ A △ B = {2, 4, 5, 7, 9, 11}

C △ A ⨯ B = {(6, 6), (5, 6), (9, 8), (10, 6), (11, 2), (6, 2), (9, 4), (7, 2), (5, 8), (10, 8), (11, 10), (7, 6), (6, 10), (10, 2), (8, 10), (7, 10), (8, 6), (6, 4), (5, 4), (11, 4), (8, 2), (10, 4), (5, 2), (9, 10), (9, 6), (5, 10), (6, 8), (10, 10), (9, 2), (11, 8), (7, 4), (8, 8), (7, 8), (11, 6), (8, 4)}

C ⨯ A ⋂ B = {(11, 2), (6, 2), (9, 4), (8, 4), (7, 2), (1, 2), (4, 4), (2, 2), (3, 4), (6, 4), (3, 2), (5, 4), (11, 4), (8, 2), (10, 4), (1, 4), (4, 2), (9, 2), (7, 4), (5, 2), (2, 4), (10, 2)}

C ⨯ A ⋃ B = {(7, 3), 2, (1, 3), 4, (9, 1), 6, 8, (11, 2), 10, (2, 1), (6, 2), (9, 4), (5, 1), (10, 3), (7, 2), (1, 2), (3, 3), (8, 1), (4, 4), (10, 2), (6, 3), (11, 1), (2, 2), (3, 4), (4, 1), (1, 1), (6, 4), (3, 2), (5, 4), (11, 4), (8, 2), (10, 4), (7, 1), (9, 3), (1, 4), (2, 3), (10, 1), (4, 2), (5, 3), (8, 3), (9, 2), (6, 1), (3, 1), (11, 3), (7, 4), (4, 3), (5, 2), (2, 4), (8, 4)}

C 

C ⋃ B \ B = {1, 3, 5, 7, 9, 11}

C ⋃ B △ B = {1, 3, 5, 7, 9, 11}

C ⋃ B ⨯ B = {(4, 8), (5, 6), (2, 8), (6, 6), (9, 8), (10, 6), (11, 2), (6, 2), (1, 6), (9, 4), (5, 8), (7, 2), (1, 2), (10, 8), (11, 10), (3, 10), (4, 4), (6, 10), (7, 6), (8, 10), (10, 2), (3, 6), (2, 2), (1, 10), (7, 10), (8, 6), (4, 10), (3, 2), (2, 6), (5, 4), (6, 4), (8, 2), (10, 4), (11, 4), (5, 2), (1, 4), (2, 10), (9, 10), (4, 2), (9, 6), (7, 8), (5, 10), (4, 6), (6, 8), (10, 10), (9, 2), (11, 8), (3, 8), (7, 4), (1, 8), (8, 8), (3, 4), (2, 4), (11, 6), (8, 4)}

C \ B ⋂ B = {1, 3, 5, 7, 9, 11}

C \ B ⋃ B = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}

C \ B \ B = {1, 3, 5, 7, 9, 11}

C \ B △ B = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}

C \ B ⨯ B = {(5, 6), (9, 8), (11, 2), (1, 6), (9, 4), (7, 2), (5, 8), (1, 2), (11, 10), (3, 10), (7, 6), (3, 6), (7, 10), (1, 10), (3, 4), (3, 2), (5, 4), (11, 4), (5, 2), (1, 4), (9, 10), (9, 6), (5, 10), (9, 2), (11, 8), (3, 8), (7, 4), (1, 8), (7, 8), (11, 6)}

C △ B ⋂ B = {1, 3, 5, 7, 9, 11

**4) Даны 4 последовательности.**  
  
Необходимо:  
1. исследовать их на монотонность;   
2. исследовать на ограниченность;  
3. найти пятый по счету член.  


**Решение выполнено "на листочке":**
https://drive.google.com/drive/folders/1bRoQ8BLFyol5FPyVII1rh0wYQKc1R-zO

**5.1)  *На языке Python предложить алгоритм вычисляющий численно предел $$\lim _{n \rightarrow+\infty} \frac{n}{\sqrt[n]{n !}}$$ с точностью $\varepsilon=10^{-7}$**


##### формула Стирлинга (также формула Муавра — Стирлинга) для аппроксимации функции факториала  
$$n ! \approx \sqrt{2 \pi n}\left(\frac{n}{e}\right)^{n}$$  

После подстановки формулы Стирлинга в функцию получаем:  
$$\lim _{n \rightarrow+\infty} \frac{e}{\sqrt[2n]{\pi n}}$$

In [12]:
def f(n):
    return (np.exp(1))/((np.pi*n)**(1/(2*n)))

Так как значения функции Стирлинга имеют приближенное значение к значению функции факториала, будем искать такие два ближайшие значения функции, разность которых меньше 1e-09

In [13]:
prev = f(1)
for i in range(2, 200000000):
    new_val = f(i)
    if new_val - prev >= 1e-09:
        prev = new_val
    else:
        print(i, new_val, prev)
        break

127163 2.7181439763031636 2.7181439753031715


За 127163 итерацию установлено что с точностью $\varepsilon=10^{-7}$ значение предела функции ~ 2.7181439763031636

**5.2) *Предложить оптимизацию алгоритма, полученного в задании 5.1, ускоряющую его сходимость.**

Для уменьшения количества итераций можно ускорить перебор значений n

In [14]:
i = 2
iters_num = 0
while True:
    iters_num += 1
    prev = f(i-1)
    new_val = f(i)
#     dif_sign = np.log10(abs(new_val - prev))
#     if dif_sign > -8:
    if new_val - prev >= 1e-09:
        i *= 10
    else:
        print(i, new_val, prev)
        print("Iters num:", iters_num)
        break

200000 2.7181911018633533 2.7181911014437032
Iters num: 6


После оптимизации подсчет значения предела занял 6 проходов по 2 подсчета каждый.

Если уменьшать значение разности на порядки, то оптимизация будет еще более выгодной.